cachelib/allocator/BackgroundMover.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) Intel and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "cachelib/allocator/BackgroundMoverStrategy.h"
+#include "cachelib/allocator/CacheStats.h"
+#include "cachelib/common/AtomicCounter.h"
+#include "cachelib/common/PeriodicWorker.h"
+
+namespace facebook {
+namespace cachelib {
+
+// wrapper that exposes the private APIs of CacheType that are specifically
+// needed for the cache api
+template <typename C>
+struct BackgroundMoverAPIWrapper {
+  static size_t traverseAndEvictItems(C& cache,
+                                      unsigned int pid,
+                                      unsigned int cid,
+                                      size_t batch) {
+    return cache.traverseAndEvictItems(pid, cid, batch);
+  }
+
+  static size_t traverseAndPromoteItems(C& cache,
+                                        unsigned int pid,
+                                        unsigned int cid,
+                                        size_t batch) {
+    return cache.traverseAndPromoteItems(pid, cid, batch);
+  }
+};
+
+enum class MoverDir { Evict = 0, Promote };
+
+// Periodic worker that evicts items from tiers in batches
+// The primary aim is to reduce insertion times for new items in the
+// cache
+template <typename CacheT>
+class BackgroundMover : public PeriodicWorker {
+ public:
+  using Cache = CacheT;
+  // @param cache               the cache interface
+  // @param strategy            the stragey class that defines how objects are
+  // moved,
+  //                            (promoted vs. evicted and how much)
+  BackgroundMover(Cache& cache,
+                  std::shared_ptr<BackgroundMoverStrategy> strategy,
+                  MoverDir direction_);
+
+  ~BackgroundMover() override;
+
+  BackgroundMoverStats getStats() const noexcept;
+  std::map<PoolId, std::map<ClassId, uint64_t>> getClassStats() const noexcept;
+
+  void setAssignedMemory(std::vector<MemoryDescriptorType>&& assignedMemory);
```

Comment:
This is only needed when we add pools right?

Instead of this, can we stop all background workers and recreate them with the new mem assignments?

Typically, applications do not add a new Pool when it is serving user requests. They typically adjust the pools on startup (e.g. right after recovering the cache from previous run).

Reply:
> This is only needed when we add pools right?

Yes, that's true.

> Instead of this, can we stop all background workers and recreate them with the new mem assignments?

We could, but then we would need to introduce extra mutex(es) in CacheAllocator class to protect `backgroundEvictor_` and `backgroundPromoter_` variables. We need to access them for example when we call `backgroundEvictor_[...]->wakeUp()`. Woul you prefer to have that extra mutex instead of the one in BackgroundMover?

Reply:
ah you're right. Actually this made me realize there's a bug in existing CacheAllocator where we try to wake up the Rebalancer after an allocation failure (so we can move slabs around sooner). This actually can crash if user is stopping/creating a new rebalancer at the same time. https://github.com/facebook/CacheLib/blob/main/cachelib/allocator/CacheAllocator-inl.h#L368

----

Let's keep this logic as-is.

----

With the logic as-is today, we still have the race where user calls addPool while they start/stop BackgroundMover worker. Can you add some comments just to clarify that it's not expected that user will call addPool and start/stop workers from two different threads?